python/pyarrow/_compute.pyx

```diff
@@ -2202,12 +2202,18 @@ def _group_by(args, keys, aggregations):
     _pack_compute_args(args, &c_args)
     _pack_compute_args(keys, &c_keys)
 
-    for aggr_func_name, aggr_opts in aggregations:
+    # reference into the flattened list of arguments for the aggregations
+    field_ref = 0
+    for aggr_arg_names, aggr_func_name, aggr_opts in aggregations:
         c_aggr.function = tobytes(aggr_func_name)
         if aggr_opts is not None:
             c_aggr.options = (<FunctionOptions?>aggr_opts).wrapped
         else:
             c_aggr.options = <shared_ptr[CFunctionOptions]>nullptr
+        for _ in aggr_arg_names:
+            c_aggr.target.push_back(CFieldRef(<int> field_ref))
+            field_ref += 1
```

Comment:
Not for this PR, but would it make sense for this code to start referring by field name instead of integer field? (but then we would need to be able to pass the `args` as an actual table/recordbatch instead of a list of arrays?)

Reply:
I'm not super familiar with the integration here but I think you're right and the Python side doesn't seem to provide named columns.

Reply:
I had simplified this in #14867 to just get rid of the field refs but I think I'll need to revisit that logic after this merges.  I think it could be made a lot simpler.

Today we are going Table -> vector<Array> -> C++ -> Table -> ExecPlan -> Table -> python

Because python is giving C++ a vector of arrays there are no "names" and named refs wouldn't make sense.  It also means that C++ has to invent names for the output columns.  Then python just renames and clobbers those invented names.

I think this would be a lot simpler as just:

Table -> C++ -> ExecPlan -> Table -> python

I'll put #14867 on hold for now and take a look again when this merges.